### PR TITLE
Fix for Order by issue in select into statement with identity function

### PR DESCRIPTION
**Pull request by deepakshi-mittal**
### Description
Fix for Order by issue in select into statement with identity function

Modified  transformSelectIntoStmt, remove unused variable queryString
Modified  SortClause and Distinct Clause of select into, removed identity function if exists in it
Added alter table add identity statement which will be executed after select into statement
Added IsA check for SortGroupClause
Removed unused variable tsql_select_into_seq_oid
Removed smallint and int functions,
Dropped function objects from upgrade scripts
Used strcasecmp inplace of strncasecmp

Added test cases for select into identity with 
- top clause 
- distinct clause
- order by clause

Task: BABEL-4388

Engine Pr:https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/247

### Issues Resolved

BABEL-4388, BABEL-539

### Test Scenarios Covered ###
* **Use case based -**

Added test cases for select into identity with 
- Different syntax IDENTITY(datatype) ,  IDENTITY(datatype, seed), IDENTITY(datatype, seed, increment)
- negative seed and increment values
- bigint, numeric, decimal datatypes etc
- top clause 
- distinct clause
- order by clause asc, desc
- order by multiple columns
- selecting into temporary table
- Self Join 

* **Negative based -**
- User defined datatypes (Not supported)
- Non numeric types like char (Not allowed)
- calling internal functions directly
- Two identity columns in a query
- IDENTITY() without specifying typename
-  IDENTITY(int) without specifying column name for it

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 1]
```diff
@@ -4518,4 +4500,35 @@ static Node* optimize_explicit_cast(ParseState *pstate, Node *node)
...
+	bool is_identity_column = false;
+
+	foreach (o_target, targetlist)
+	{
+		ResTarget *res = (ResTarget *)lfirst(o_target);
+		is_identity_column = false;
+		if (IsA(res->val, FuncCall))
+		{
+			FuncCall *fc = (FuncCall *)res->val;
+			if (strncasecmp(strVal(llast(fc->funcname)), "identity_into", strlen("identity_into")) == 0)
```
Comment (xhfanhe) [inline contrib/babelfishpg_tsql/src/hooks.c]: What if funcname is "identity_into_xxx", will this still pass? I think strcmp is sufficient enough.

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 2]
```diff
@@ -4518,4 +4500,35 @@ static Node* optimize_explicit_cast(ParseState *pstate, Node *node)
...
+bbf_rewrite_targetlist(List *targetlist)
+{
+	ListCell *o_target;
+	List *modifiedTargetList = NIL;
+	ResTarget *identity_column_res = NULL;
+	bool is_identity_column = false;
+
+	foreach (o_target, targetlist)
+	{
+		ResTarget *res = (ResTarget *)lfirst(o_target);
```
Comment (xhfanhe) [inline contrib/babelfishpg_tsql/src/hooks.c]: Need to first check if this Node is a ResTarget

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c [hunk 3]
```diff
@@ -233,8 +233,35 @@ TsqlFunctionConvert(TypeName *typename, Node *arg, Node *style, bool try, int lo
 Node *
 TsqlFunctionIdentityInto(TypeName *typename, Node *seed, Node *increment, int location)
```
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:234]: will identity function be called in this case ?

Comment (deepakshi-mittal) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:234]: The actual functions wont be invoked, but we need the typename, seed and increment value. I can make a single function

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 4]
```diff
@@ -5743,33 +5744,64 @@ static List *transformSelectIntoStmt(CreateTableAsStmt *stmt, const char *queryS
...
-				altseqstmt->for_identity = true;
+		if (seen_identity)
+		{
+			if (q->sortClause)
+			{
+				List *modifiedSortList = NIL;
+				ListCell *olitem;
+				foreach (olitem, q->sortClause)
+				{
+					SortGroupClause *sortcl = (SortGroupClause *)lfirst(olitem);
```
Comment (xhfanhe) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Always check IsA() to make sure we are getting the correct struct

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 5]
```diff
@@ -5663,49 +5666,49 @@ static int64 get_identity_into_args(Node *node)
...
+		AttrNumber current_resno = 0;
+		Index identity_ressortgroupref = 0;
+		List *modifiedTargetList = NIL;
+
 		foreach (elements, q->targetList)
 		{
 			TargetEntry *tle = (TargetEntry *)lfirst(elements);
 
-			if (tle->expr && IsA(tle->expr, FuncExpr) && strncasecmp(get_func_name(((FuncExpr *)(tle->expr))->funcid), "identity_into", strlen( "identity_into")) == 0)
+			if (tle->expr && IsA(tle->expr, FuncExpr) && strncasecmp(get_func_name(((FuncExpr *)(tle->expr))->funcid), "identity_into", strlen("identity_into")) == 0)
```
Comment (xhfanhe) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Same doubt for strncasecmp here

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c (hunk 3 above)
Comment (deepakshi-mittal) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:234]: Removed two extra function int and smallint and just kep one, bigint which will take care of all numeric cases

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 1 above)
Comment (deepakshi-mittal) [reply contrib/babelfishpg_tsql/src/hooks.c]: Changed to strcasecmp "identity_into_bigint", Removed two extra function int and smallint and just kep one, bigint which will take care of all numeric case

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 2 above)
Comment (deepakshi-mittal) [reply contrib/babelfishpg_tsql/src/hooks.c]: Removed this hook as the code works without modifying target list

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 4 above)
Comment (deepakshi-mittal) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: Done

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 5 above)
Comment (deepakshi-mittal) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: changed to strcasecmp

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 6]
```diff
@@ -2312,34 +2311,22 @@ bbf_set_context_info(PG_FUNCTION_ARGS)
 	PG_RETURN_VOID();
 }
 
-/**
- * Common function to get next value of sequence for Identity function in Select Into
- */
-static int64 
-get_identity_next_value(void)
```
Comment (deepakshi-mittal) [inline contrib/babelfishpg_tsql/runtime/functions.c:2319]: Its no longer needed as we are using alter table add identity statement now

**Review by xhfanhe — APPROVED** [review]

**Review by forestkeeper — APPROVED** [review]

**Pull request merged**